tests/org.dataflowanalysis.analysis.tests/src/org/dataflowanalysis/analysis/tests/dfd/MicroSecEndTest.java

```diff
@@ -0,0 +1,296 @@
+            violationsSet.add(4);
+        }
+    }
+
+    // Authentication tokens should be validated
+    private void hasTokenValidation(Set<Integer> violationsSet, int variant, AbstractVertex<?> node) {
+        if (hasDataCharacteristicAcrossVariables(node, "Stereotype", "entrypoint") && hasNodeCharacteristic(node, "Stereotype", "internal")
+                && !hasNodeCharacteristic(node, "Stereotype", "token_validation")) {
+            violationsSet.add(5);
+        }
+    }
+
+    // A limit for the maximum number of login attempts before preventive measures are taken should exist.
+    private void hasLoginAttemptsRegulation(Set<Integer> violationsSet, int variant, AbstractVertex<?> node) {
+        if (hasNodeCharacteristic(node, "Stereotype", "authorization_server")
+                && !hasNodeCharacteristic(node, "Stereotype", "login_attempts_regulation")) {
+            violationsSet.add(6);
+
+        }
+    }
+
+    // All communication traffic from external users and entities should be encrypted using secure communication
+    // protocols
+    private void hasEncryptedEntryConnection(Set<Integer> violationsSet, int variant, AbstractVertex<?> node) {
+        if (hasDataCharecteristicViolation(node, "Stereotype", "entrypoint", "Stereotype", "encrypted_connection")) {
+            violationsSet.add(7);
+        }
+    }
+
+    // All communication between the services should be encrypted using secure communication protocols.
+    private void hasEncrytedInternalConnection(Set<Integer> violationsSet, int variant, AbstractVertex<?> node) {
+        if (hasDataCharecteristicViolation(node, "Stereotype", "internal", "Stereotype", "encrypted_connection")) {
+            violationsSet.add(8);
+        }
+    }
+
+    // A central logging subsystem which includes a monitoring dashboard should exist.
+    private void hasLoggingServer(AbstractTransposeFlowGraph aTFG, Map<Integer, List<AbstractTransposeFlowGraph>> existenceViolations) {
+        existenceViolations.putIfAbsent(9, new ArrayList<AbstractTransposeFlowGraph>());
+        if (!hasNodeWithCharacteristic(aTFG, "Stereotype", "logging_server")) {
+            existenceViolations.get(9)
+                    .add(aTFG);
+        }
+    }
+
+    // For all microservices, there should exist a local logging agent decoupled from the microservice but deployed
+    // on the same host. Log data from microservices should not be send to the central logging system directly, but
+    // collected by the logging agent, written to a local file, and eventually send to the central system by it.
+    private void hasLocalLogging(Set<Integer> violationsSet, int variant, AbstractVertex<?> node) {
+        if (hasNodeCharacteristic(node, "Stereotype", "internal") && !hasNodeCharacteristic(node, "Stereotype", "local_logging")) {
+            violationsSet.add(10);
+            violationsSet.add(11);
```

Comment:
No logging agent always also violates the sanitization?

Reply:
Yesish, if there is no logging 11 is still violated, since it build on top of 10

Reply:
Then please add a small comment that states exactly this